search_params.go

```diff
@@ -0,0 +1,103 @@
+package faiss
+
+/*
+#include <faiss/c_api/Index_c.h>
+#include <faiss/c_api/IndexIVF_c.h>
+#include <faiss/c_api/impl/AuxIndexStructures_c.h>
+*/
+import "C"
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+)
+
+type SearchParams struct {
+	sp *C.FaissSearchParameters
+}
+
+// Delete frees the memory associated with s.
+func (s *SearchParams) Delete() {
+	if s == nil || s.sp == nil {
+		return
+	}
+	C.faiss_SearchParameters_free(s.sp)
+}
+
+type searchParamsIVF struct {
+	NprobePct   float32 `json:"nprobe_pct,omitempty"`
+	MaxCodesPct float32 `json:"max_codes_pct,omitempty"`
+}
+
+func (s *searchParamsIVF) Validate() error {
+	if s.NprobePct < 0 || s.NprobePct > 100 {
+		return fmt.Errorf("invalid IVF search params, nprobe_pct:%v, should be in range [0, 100]", s.NprobePct)
+	}
+
+	if s.MaxCodesPct < 0 || s.MaxCodesPct > 100 {
+		return fmt.Errorf("invalid IVF search params, max_codes_pct:%v, should be in range [0, 100]", s.MaxCodesPct)
+	}
+
+	return nil
+}
+
+// Attempt to create IVF search params
+// and fall back to default search params if the index is not IVF.
+//
+// Always return a valid SearchParams object,
+// thus caller must clean up the object
+// by invoking Delete() method, even if an error is returned.
+func NewSearchParamsIVF(idx Index, params json.RawMessage, sel *C.FaissIDSelector,
+) (*SearchParams, error) {
+
+	rv := &SearchParams{}
+	if c := C.faiss_SearchParameters_new(&rv.sp, sel); c != 0 {
+		return rv, fmt.Errorf("failed to create faiss search params")
+	}
+
+	// # check if the index is IVF and set the search params
+
+	i := C.faiss_IndexIVF_cast(idx.cPtr())
+	if i == nil { // not an IVF index, return the default search params
+		return rv, nil
+	}
+
+	rv.sp = C.faiss_SearchParametersIVF_cast(rv.sp)
+	if len(params) == 0 {
+		return rv, nil
+	}
+
+	var ivf searchParamsIVF
+	if err := json.Unmarshal(params, &ivf); err != nil {
+		return rv, fmt.Errorf("failed to unmarshal IVF search params, err:%v", err)
+	}
+	if err := ivf.Validate(); err != nil {
+		return rv, err
+	}
+
+	var nprobe, maxCodes int
+
+	if ivf.NprobePct > 0 {
+		nlist := C.faiss_IndexIVF_nlist(i)
+		nprobe = int(float32(nlist) * (ivf.NprobePct / 100))
+	} else {
+		idxNprobe := C.faiss_IndexIVF_nprobe(i)
```

Comment:
Is this else necessary to do? Wouldn't this nprobe be automatically picked in faiss otherwise?

Reply:
I think this is necessary, otherwise faiss will use nprobe=1 

Whenever we are searching with parameters we have to explicity provide the value of nprobe (be it user supplied or be it what we set at index creation), else it will pick the default (which is nprobe=1)

Reply:
Please confirm and add some commentary indicating this then.
Also line 84 is not really necessary, 85 can be adapted to ..
```
        nprobe = int(C.faiss_IndexIVF_nprobe(i))
```

Reply:
Done
Thanks